src/deluge/gui/views/automation_instrument_clip_view.cpp

```diff
@@ -0,0 +1,3138 @@
+				instrumentClipView.editPadPresses[i].yDisplay = yDisplay;
+				instrumentClipView.editPadPresses[i].xDisplay = xDisplay;
+				instrumentClipView.numEditPadPresses++;
+				instrumentClipView.numEditPadPressesPerNoteRowOnScreen[yDisplay]++;
+				enterUIMode(UI_MODE_NOTES_PRESSED);
+			}
+		}
+	}
+
+	// Or if pad press ended...
+	else {
+
+		// Find the corresponding press, if there is one
+		int32_t i;
+		for (i = 0; i < kEditPadPressBufferSize; i++) {
+			if (instrumentClipView.editPadPresses[i].isActive
+			    && instrumentClipView.editPadPresses[i].yDisplay == yDisplay
+			    && instrumentClipView.editPadPresses[i].xDisplay == xDisplay) {
+				break;
+			}
+		}
+
+		// If we found it...
+		if (i < kEditPadPressBufferSize) {
+			instrumentClipView.endEditPadPress(i);
+
+			instrumentClipView.checkIfAllEditPadPressesEnded();
+		}
+	}
+}
+
+//audition pad action
+//don't think I changed anything in here compared to the instrument clip view version
+
+void AutomationInstrumentClipView::auditionPadAction(int32_t velocity, int32_t yDisplay, bool shiftButtonDown) {
+
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+	ModelStack* modelStack = setupModelStackWithSong(modelStackMemory, currentSong);
+
+	bool clipIsActiveOnInstrument = makeCurrentClipActiveOnInstrumentIfPossible(modelStack);
+
+	InstrumentClip* clip = getCurrentClip();
+	Instrument* instrument = (Instrument*)clip->output;
+
+	bool isKit = (instrument->type == InstrumentType::KIT);
+
+	ModelStackWithTimelineCounter* modelStackWithTimelineCounter = modelStack->addTimelineCounter(clip);
+
+	ModelStackWithNoteRow* modelStackWithNoteRowOnCurrentClip =
+	    clip->getNoteRowOnScreen(yDisplay, modelStackWithTimelineCounter);
+
+	Drum* drum = NULL;
+
+	// If Kit...
+	if (isKit) {
+
+		if (modelStackWithNoteRowOnCurrentClip->getNoteRowAllowNull()) {
+			drum = modelStackWithNoteRowOnCurrentClip->getNoteRow()->drum;
+			if (((Kit*)instrument)->selectedDrum != drum) {
+				if (!instrumentClipView.getAffectEntire()) {
+					initParameterSelection();
+				}
+			}
+		}
+
+		// If NoteRow doesn't exist here, we'll see about creating one
+		else {
+
+			// But not if we're actually not on this screen
+			if (getCurrentUI() != this) {
+				return;
+			}
+
+			// Press-down
+			if (velocity) {
+
+				instrumentClipView.setSelectedDrum(NULL);
+			}
+
+			goto getOut;
+		}
+	}
+
+	// Or if synth
+	else if (instrument->type == InstrumentType::SYNTH) {
+		if (velocity) {
+			if (getCurrentUI() == &soundEditor && soundEditor.getCurrentMenuItem() == &menu_item::multiRangeMenu) {
+				menu_item::multiRangeMenu.noteOnToChangeRange(clip->getYNoteFromYDisplay(yDisplay, currentSong)
+				                                              + ((SoundInstrument*)instrument)->transpose);
+			}
+		}
+	}
+
+	// Recording - only allowed if currentClip is activeClip
+	if (clipIsActiveOnInstrument && playbackHandler.shouldRecordNotesNow() && currentSong->isClipActive(clip)) {
```

Comment:
I guess this is required to support recording when the automation view is open but it would be nice to kick this code somewhere it can be shared, especially if everything is just duplicated from InstrumentClipView -- How does the keyboard view handle it?

Reply:
Keyboard view calls auditionPadAction in instrumentClipView directly 

most of the code does come from instrument clip view, but I did strip out of some stuff (audition pad actions I don't want in my view) and added a few line items. Would it be better to call instrument clip view directly and put an if statement around the code I don't want executed in my view (e.g. if getCurrentUI() != &automationInstrumentClipView)

Reply:
This is fine for now. It sounds like we really need to move a bunch of logic from `instrumentClipView` to `ClipMinder` or something like that so it can be shared more cleanly, but that's a follow-up problem.